### ISSUE TITLE
Relative path support for userAuthorizationUri and accessTokenUri

### ISSUE THREAD

**Issue by kakawait** [spring-projects/spring-security-oauth#671]
On my project we have an *UAA* on same domain than *OAuth2 Sso* (via `@EnableOauth2Sso`).

Moreover to easily deploy application on every possible environment we try to do not have any absolute URLs.

Thus I want to know if is possible and not *a no-sense* to be able to configure *OAuth2* client properties:

- `security.oauth2.client.accessTokenUri`
- `security.oauth2.client.userAuthorizationUri`

as relative *path*.

I don't know how, possible solutions:

- new properties (ugly version).
- create an Uri builder/interceptor that can be override and instead of calling `resource.getAccessTokenUri()` call builder/interceptor instead.
- other?

**Comment by kakawait** [comment]
If you think my point of view is correct I can create `PR`

**Comment by kakawait** [comment]
Today to avoid using *hardcode* hostname I have to do following modification:

1. Rewrite `OAuth2ClientContextFilter` to support `URI` in favor to `URL`! By supporting `URI` you can now use relative path (supported by header `Location` and browser)

```
userAuthorizationUri: /uaa/oauth/authorize
```

but for `accessTokenUri` I just have a trick by setting following thing on Zuul 

```
accessTokenUri: http://localhost:${server.port}/uaa/oauth/token
```

Thus `Oauth2RestTemplate` will use Zuul to call AuthorizationService. It could be better to add `LoadBalancerd` capability to `Oauth2RestTemplate` for replacing following tricks by

```
accessTokenUri: http://authorization-service/uaa/oauth/token
```

**Comment by dsyer** [comment]
I'm not sure we can handle any of the zuul or load balancer features here, so maybe you could move the discussion to spring-cloud/spring-cloud-security?

**Comment by kakawait** [comment]
@dsyer There is a way to transfer issue between project?

**Comment by dsyer** [comment]
No, but you can paste a link in the other project's issue and it will show up here automatically.

**Comment by dsyer** [comment]
Let's leave this one open to cover the changes you are asking for in `OAuth2ClientContextFilter` (to support relative URLs). The token endpoint would have to be covered in the client, so I don't think it makes sense here (but might in a Spring Cloud setting).

**Comment by kakawait** [comment]
But should I open an issue on spring-cloud/spring-cloud-security too? Maybe I will wait you completely finish analysis of https://github.com/kakawait/uaa-behind-zuul-sample before opening something new

**Comment by dsyer** [comment]
> But should I open an issue on spring-cloud/spring-cloud-security too?

Yes, please, I think that's where the change would have to be made in Spring Cloud.

**Comment by jgrandja** [comment]
@dsyer Is there a change required here? Or is it just in Spring Cloud? Given that the discussion started with `@EnableOauth2Sso`.

**Comment by dsyer** [comment]
Yes, some change would be here since it relates to the client filter. (And `@EnableOAuth2Sso` is a Spring Boot feature now.)

**Comment by rwinch** [comment]
This might be resolved by using https://jira.spring.io/browse/SPR-13614

**Comment by hsjankie** [comment]
@dsyer Was the relative change added yet, it would help a lot in our setup.

**Comment by dsyer** [comment]
There's no change here. If there was you would see the status updated.

**Comment by kakawait** [comment]
@hsjankie @jgrandja I found a solution for `accessTokenUri` by simply enhance the used `RestTemplate` with `Ribbon`, please check here https://github.com/spring-cloud/spring-cloud-security/issues/61#issuecomment-287376098
